taskcat/testing/manager.py

```diff
@@ -0,0 +1,252 @@
+        )
+        self.test_definition.create_stacks()
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        terminal_printer.report_test_progress(stacker=self.test_definition)
+
+    def end(
+        self,
+        no_delete: bool = False,
+        keep_failed: bool = False,
+        dont_wait_for_delete: bool = True,
+    ):
+        """
+        docstring
+        """
+
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        status = self.test_definition.status()
+
+        # Delete Stacks
+        if no_delete:
+            LOG.info("Skipping delete due to cli argument")
+        elif keep_failed:
+            if len(status["COMPLETE"]) > 0:
+                LOG.info("deleting successful stacks")
+                self.test_definition.delete_stacks({"status": "CREATE_COMPLETE"})
+        else:
+            self.test_definition.delete_stacks()
+
+        if not dont_wait_for_delete:
+            terminal_printer.report_test_progress(stacker=self.test_definition)
+
+        # TODO: summarise stack statusses (did they complete/delete ok) and print any
+        #  error events
+
+        # Delete Templates and Buckets
+        buckets = self.config.get_buckets()
+
+        if not no_delete or (keep_failed is True and len(status["FAILED"]) == 0):
+            deleted: ListType[str] = []
+            for test in buckets.values():
+                for bucket in test.values():
+                    if (bucket.name not in deleted) and not bucket.regional_buckets:
+                        bucket.delete(delete_objects=True)
+                        deleted.append(bucket.name)
+
+        # 9. raise if something failed
+        # - grabbing the status again to ensure everything deleted OK.
+
+        status = self.test_definition.status()
+        if len(status["FAILED"]) > 0:
+            raise TaskCatException(
+                f'One or more stacks failed tests: {status["FAILED"]}'
+            )
+
+    def report(
+        self, output_directory: str = "./taskcat_outputs",
+    ):
+        """
+        Generates a report of the status of Cloudformation stacks.
+        """
+        report_path = Path(output_directory).resolve()
+        report_path.mkdir(exist_ok=True)
+        cfn_logs = _CfnLogTools()
+        cfn_logs.createcfnlogs(self.test_definition, report_path)
+        ReportBuilder(
+            self.test_definition, report_path / "index.html"
+        ).generate_report()
+
+
+def _trim_regions(regions, config):
+    if regions != "ALL":
+        for test in config.config.tests.values():
+            to_pop = []
+            idx = 0
+            if test.regions:
+                for _ in test.regions:
+                    if test.regions[idx] not in regions.split(","):
+                        to_pop.append(idx)
+                    idx += 1
+                to_pop.reverse()
+                for idx in to_pop:
+                    test.regions.pop(idx)
```

Comment:
as I mentioned earlier it seemed like this kinda thing could probably be handled in the config object since we already pass it the regions.

Reply:
Agreed. This and trim-tests were added relatively quickly post `0.9.0`.